File: DTO/Invoice/TravelMatrix.cs
using System;
using System.Collections.Generic;
using System.Text;
using DTO.CommonClass;

namespace DTO.Invoice
{
    public class TravelMatrix
    {
        public int? Id { get; set; }
        public int? CustomerId { get; set; }
        public int TariffTypeId { get; set; }
        public int CountryId { get; set; }
        public int ProvinceId { get; set; }
        public int? CountyId { get; set; }
        public int? CityId { get; set; }
        public int? InspPortId { get; set; }
        public int? InspPortCityId { get; set; }
        public double? DistanceKM { get; set; }
        public double? TravelTime { get; set; }
        public double? BusCost { get; set; }
        public double? TrainCost { get; set; }
        public double? TaxiCost { get; set; }
        public double? HotelCost { get; set; }
        public double? OtherCost { get; set; }
        public double? MarkUpCost { get; set; }
        public double? AirCost { get; set; }
        public double? MarkUpCostAir { get; set; }
        public double FixExchangeRate { get; set; }
        public string Remarks { get; set; }
        public int? SourceCurrencyId { get; set; }
        public int? TravelCurrencyId { get; set; }
    }
    public class TravelMatrixSearch
    {
        public int? Id { get; set; }
        public int? CustomerId { get; set; }
        public int? TariffTypeId { get; set; }
        public int? CountryId { get; set; }
        public int? ProvinceId { get; set; }
        public int? CountyId { get; set; }
        public int? CityId { get; set; }
        public int? CountyCityId { get; set; }
        public int? InspPortId { get; set; }
        public double? DistanceKM { get; set; }
        public double? TravelTime { get; set; }
        public double? BusCost { get; set; }
        public double? TrainCost { get; set; }
        public double? TaxiCost { get; set; }
        public double? HotelCost { get; set; }
        public double? OtherCost { get; set; }
        public double? MarkUpCost { get; set; }
        public double? AirCost { get; set; }
        public double? MarkUpCostAir { get; set; }
        public double? FixExchangeRate { get; set; }
        public string Remarks { get; set; }
        public int? SourceCurrencyId { get; set; }
        public int? TravelCurrencyId { get; set; }

        public string CustomerName { get; set; }
        public string TariffTypeName { get; set; }
        public string CountryName { get; set; }
        public string ProvinceName { get; set; }
        public string CityName { get; set; }
        public string CountyName { get; set; }
        public string InspPortName { get; set; }
        public string InspPortCityName { get; set; }
        public string SourceCurrencyName { get; set; }
        public string TravelCurrencyName { get; set; }

        public int? InspPortCityId { get; set; }

       

    }

    public enum TravelMatrixConfig
    {
        Province=1,
        City=2,
        County=3
    }

    public enum TravelMatrixResponseResult
    {
        Success = 1,
        Error = 2,
        RequestNotCorrectFormat = 3,
        NotFound = 4,
        DefaultData = 5,
        MoreMatrixExists = 6,
        PriceCardNotCorrect = 7
    }

    public class TravelMatixSaveResponse
    {
        public TravelMatrixResponseResult Result { get; set; }
        public IEnumerable<TravelMatrixExists> ExistsData { get; set; }
    }

    public class TravelMatrixSearchResponse
    {
        public IEnumerable<TravelMatrixSearch> GetData { get; set; }
        public TravelMatrixResponseResult Result { get; set; }

        public int TotalCount { get; set; }

        public int Index { get; set; }

        public int PageSize { get; set; }

        public int PageCount { get; set; }

        public bool IsDataExist { get; set; }
    }

    public class TravelMatrixSummary
    {
        public int? TariffTypeId { get; set; }
        public int? TravelCurrencyId { get; set; }
        public int? CustomerId { get; set; }
        public int? SourceCurrencyId { get; set; }
        public int? CountryId { get; set; }
        public int? ProvinceId { get; set; }
        public int? CityId { get; set; }
        public int? CountyId { get; set; }

        public int? Index { get; set; }

        public int? PageSize { get; set; }

        public bool IsExport { get; set; }

        public int SearchTypeId { get; set; }
    }

    public class AreaDetails
    {
        public int CountryId { get; set; }
        public int ProvinceId { get; set; }
        public int CityId { get; set; }

        public string CountryName { get; set; }
        public string ProvinceName { get; set; }
        public string CityName { get; set; }

        public int CountyId { get; set; }

        public string AreaCountryName { get; set; }
        public string AreaProvinceName { get; set; }
        public string AreaCityName { get; set; }
        public string AreaCountyName { get; set; }
    }
    public class AreaData
    {
        public int Id { get; set; }
        public List<CommonDataSource> DataList { get; set; }
    }
    public class AreaDataResponse
    {
        public TravelMatrixResponseResult Result { get; set; }
        public List<AreaData> AreaDataList { get; set; }
    }
    public class TravelMatixDeleteResponse
    {
        public TravelMatrixResponseResult Result { get; set; }
    }

    public class TravelMatrixExists
    {
        public string CustomerName { get; set; }
        public string TariffTypeName { get; set; }
        public string CountryName { get; set; }
        public string CountyName { get; set; }
        public string ProvinceName { get; set; }
        public string CityName { get; set; }
        public string SourceCurrencyName { get; set; }
        public string TravelCurrencyName { get; set; }
    }

    public class TravelMatrixData
    {
        public string CustomerName { get; set; }
        public string TariffTypeName { get; set; }
        public string CountryName { get; set; }
        public string CountyName { get; set; }
        public string ProvinceName { get; set; }
        public string CityName { get; set; }
        public string SourceCurrencyName { get; set; }
        public string TravelCurrencyName { get; set; }
        public int? CustomerId { get; set; }
        public int TariffTypeId { get; set; }
        public int CountryId { get; set; }
        public int ProvinceId { get; set; }
        public int? CityId { get; set; }
        public int? CountyId { get; set; }
        public int? SourceCurrencyId { get; set; }
        public int Id { get; set; }
    }

    public class ExistsTravelMatrix
    {
        public IEnumerable<TravelMatrix> RequestModel { get; set; }
        public IEnumerable<TravelMatrixData> AvailableTravelMatrix { get; set; }
    }

    public class ExistsTravelMatrixRequest
    {
        public IEnumerable<int?> SourceCurrencyList { get; set; }
        public IEnumerable<int> CountryList { get; set; }
        public IEnumerable<int> ProvinceList { get; set; }
        public IEnumerable<int?> CityList { get; set; }
        public IEnumerable<int?> CountyList { get; set; }
        public IEnumerable<int> TariffTypeList { get; set; }
    }

    public class TravelMatrixExportSummary
    {
        public string CustomerName { get; set; }
        public string TariffTypeName { get; set; }
        public string CountryName { get; set; }
        public int? CountryId { get; set; }
        public string ProvinceName { get; set; }
        public int? ProvinceId { get; set; }
        public string CityName { get; set; }
        public int? CityId { get; set; }
        public string CountyName { get; set; }
        public int? CountyId { get; set; }
        public string InspPortNameCounty { get; set; }
        public int? InspPortCountyId { get; set; }
        public string InspPortNameCity { get; set; }
        public int? InspPortCityId { get; set; }
        public string SourceCurrencyName { get; set; }
        public string TravelCurrencyName { get; set; }

        public double? DistanceKM { get; set; }
        public double? TravelTime { get; set; }
        public double? BusCost { get; set; }
        public double? TrainCost { get; set; }
        public double? TaxiCost { get; set; }
        public double? HotelCost { get; set; }
        public double? OtherCost { get; set; }
        public double? MarkUpCost { get; set; }
        public double? AirCost { get; set; }
        public double? MarkUpCostAir { get; set; }
        public double? FixExchangeRate { get; set; }
        public string Remarks { get; set; }
    }
    public class TravelMatrixRequest
    {
        public int BookingId { get; set; }
        public int? RuleId { get; set; }
        public int QuotationId { get; set; }
        public int BillMethodId { get; set; }
        public int BillPaidById { get; set; }
        public int? CurrencyId { get; set; }
        public int? CountyId { get; set; }
        public int? CityId { get; set; }
        public int? ProvinceId { get; set; }
        public int customerId { get; set; }
    }

    public class QuotationTravelMatrixResponse
    {
        public TravelMatrixSearch TravelMatrixDetails { get; set; }
        public TravelMatrixResponseResult Result { get; set; }
    }
    public class QuotationMatrixRequest
    {
        public int? CountyId { get; set; }
        public int CurrencyId { get; set; }
        public int? MatrixTypeId { get; set; }
        public int? customerId { get; set; }
        public int? CityId { get; set; }
        public int? ProvinceId { get; set; }
    }

    public class TravelMatrixSearchRequest
    {
        public int? CountyId { get; set; }
        public int? CityId { get; set; }
        public int? ProvinceId { get; set; }
    }

    public enum TravelMatrixSearchEnum
    {
        City = 1,
        County = 2
        
    }



}

File: LINK_UI/ClientApp/src/app/_Models/invoice/travelmatrix.ts
import { CityDataSourceRequest, DataSource } from "../common/common.model";
import { summaryModel } from "../summary.model";
import { Observable } from "rxjs/internal/Observable";
import { Subject } from "rxjs";

export class TravelMatrixMasterModel {
    tariffTypeList: Array<DataSource>;
    customerList: Array<DataSource>;
    travelCurrencyList: Array<DataSource>;
    sourceCurrencyList: Array<DataSource>;
    countryList: Array<DataSource>;
    provinceList: Array<DataSource>;
    cityList: Array<DataSource>;
    countyList: Array<DataSource>;

    exportDataLoading: boolean;
    tariffTypeLoading: boolean;
    travelCurrencyLoading: boolean;
    customerLoading: boolean;
    sourceCurrencyLoading: boolean;
    countryLoading: boolean;
    provinceLoading: boolean;
    cityLoading: boolean;
    countyLoading: boolean;
    searchLoading: boolean;
    saveLoading: boolean;
    addLoading: boolean;

    copyLoading: boolean;
    deleteLoading: boolean;
    selectedAllCheckBox: boolean;
    showButton: boolean;
}

export class TravelMatrixSummary extends summaryModel {
    tariffTypeId: number;
    travelCurrencyId: number;
    customerId: number;
    sourceCurrencyId: number;
    countryId: number;
    provinceId: number;
    cityId: number;
    countyId: number;
    isExport: boolean;
    searchTypeId: number;
}

export class TravelMatrixModel {

    constructor() {
        this.typeInspPortListInput = new Subject<string>();
        this.typeInspPortCityListInput = new Subject<string>();
        this.cityRequest = new CityDataSourceRequest();
    }

    id: number;
    customerLoading: boolean;
    tariffTypeLoading: boolean;
    countryLoading: boolean;
    provinceLoading: boolean;
    cityLoading: boolean;
    countyLoading: boolean;
    inspPortLoading: boolean;
    travelCurrencyLoading: boolean;
    sourceCurrencyLoading: boolean;
    isChecked: boolean;

    customerId: number;
    tariffTypeId: number;
    countryId: number;
    provinceId: number;
    countyId: number;
    cityId: number;
    inspPortId: number;
    distanceKM: number;
    travelTime: number;
    busCost: number;
    trainCost: number;
    taxiCost: number;
    hotelCost: number;
    otherCost: number;
    markUpCost: number;
    airCost: number;
    markUpCostAir: number;
    fixExchangeRate: number;
    remarks: string;
    sourceCurrencyId: number;
    travelCurrencyId: number;
    typeInspPortListInput: Subject<string>;
    provinceList: Array<DataSource>;
    countyList: Array<DataSource>;
    cityList: Array<DataSource>;
    inspPortList: Observable<DataSource[]>;
    selectedPortList: DataSource;

    typeInspPortCityListInput: Subject<string>;
    inspPortCityList: Observable<DataSource[]>;
    inspPortCityLoading: boolean;
    cityRequest: CityDataSourceRequest;
    selectedPortCityList: DataSource;
    inspPortCityId: number;
}

export interface ProvinceList {
    countryId: Number;
    provinceList: Array<DataSource>;
}

export interface CityList {
    provinceId: number;
    cityList: Array<DataSource>;
}
export interface CountyList {
    cityId: number;
    countyList: Array<DataSource>;
}

export enum TravelMatrixResponseResult {
    Success = 1,
    Error = 2,
    RequestNotCorrectFormat = 3,
    NotFound = 4,
    DefaultData = 5,
    MoreMatrixExists = 6,
    PriceCardNotCorrect = 7
}

export class TravelMatixSaveResponse {
    result: TravelMatrixResponseResult;
    existsData: Array<TravelMatrixExists>;
}

export class TravelMatixSearchResponse {
    result: TravelMatrixResponseResult;
    getData: Array<TravelMatrixModel>;
    isDataExist: boolean;
}

export class AreaData {
    id: number;
    dataList: Array<DataSource>;
    constructor() {
        this.dataList = Array<DataSource>();
    }
}

export class TravelMatrixExists {
    customerName: string;
    tariffTypeName: string;
    countryName: string;
    countyName: string;
    provinceName: string;
    cityName: string;
    sourceCurrencyName: string;
    travelCurrencyName: string;
}

export enum TariffType {
    standardA = 1,
    standardB = 2,
    customized = 3
}

export enum AreaType {
    country = 1,
    province = 2,
    city = 3
}

export class QuotationTravelMatrix {
    id: number;

    customerId: number;
    tariffTypeId: number;
    countryId: number;
    provinceId: number;
    countyId: number;
    cityId: number;
    inspPortId: number;
    distanceKM: number;
    travelTime: number;
    busCost: number;
    trainCost: number;
    taxiCost: number;
    hotelCost: number;
    otherCost: number;
    markUpCost: number;
    airCost: number;
    markUpCostAir: number;
    fixExchangeRate: number;
    remarks: string;
    sourceCurrencyId: number;
    travelCurrencyId: number;

    busMarkUp: string;
    busTotal: string;
    trainMarkUp: string;
    trainTotal: string;
    taxiMarkUp: string;
    taxiTotal: string;

    landTravelCurrencyTotal: string;
    landSourceCurrencyTotal: string;

    airMarkUp: string;
    airTotal: string;

    hotelMarkUp: string;
    hotelTotal: string;

    otherMarkUp: string;
    otherTotal: string;

    grandTotal: string;
}

export class QuotationTravelMatrixResponse {
    travelMatrixDetails: QuotationTravelMatrix;
    result: TravelMatrixResponseResult;
}

export class TravelMatrixRequest {
    countyId: number;
    cityId: number;
    provinceId: number;
    billMethodId: number;
    billPaidById: number;
    currencyId: number;
    bookingId: number;
    customerId: number;
    ruleId: number;
    quotationId: number;
}
